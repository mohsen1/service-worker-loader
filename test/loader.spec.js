import path from 'path';
import webpack from 'webpack';
import rimraf from 'rimraf';
import compile, {
	fs,
	pathToArtifacts
} from './compile';

describe('service-worker-loader', () => {

	beforeEach((done) => {
		rimraf(
			path.join(pathToArtifacts, '**', '*'),
			fs,
			done
		);
	});

	it('should emit bundle and service worker', async () => {

		const stats = await compile('serviceWorkerLoader.js', {});
		const {
			source: serviceWorkerLoaderSource
		} = stats.modules[0].modules[1];
		const {
			source: serviceWorkerSource
		} = stats.children[0].modules[0];

		expect(serviceWorkerLoaderSource).toMatchSnapshot();
		expect(serviceWorkerSource).toMatchSnapshot();
		expect(
			fs.statSync(
				path.join(pathToArtifacts, 'serviceWorker.js')
			).isFile()
		).toBe(true);
	});

	it('should emit service worker to output path', async () => {

		const outputPath = path.join(pathToArtifacts, 'sw');

		fs.mkdirSync(outputPath);

		await compile('serviceWorkerLoader.js', {
			outputPath
		});

		expect(
			fs.statSync(
				path.join(outputPath, 'serviceWorker.js')
			).isFile()
		).toBe(true);
	});

	it('should work with `HotModuleReplacementPlugin`', async () => {

		const stats = await compile('serviceWorkerLoader.js', {
			devMode: true,
			plugins: [
				new webpack.HotModuleReplacementPlugin()
			]
		});

		console.log(stats);

		expect(
			fs.statSync(
				path.join(pathToArtifacts, 'serviceWorker.js')
			).isFile()
		).toBe(true);
	});
});
