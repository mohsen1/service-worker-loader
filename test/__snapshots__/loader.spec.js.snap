// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`service-worker-loader should emit bundle and service worker 1`] = `
"export var ServiceWorkerNoSupportError = (function() {

	function ServiceWorkerNoSupportError() {
		var err = Error.call(this, 'ServiceWorker is not supported.');
		Object.setPrototypeOf(err, ServiceWorkerNoSupportError.prototype);
		return err;
	}

	ServiceWorkerNoSupportError.prototype = Object.create(Error.prototype);

	return ServiceWorkerNoSupportError;
})();

export var scriptUrl = __webpack_public_path__ + \\"serviceWorker.js\\";

export default function registerServiceWorkerIfSupported(mapScriptUrlOrOptions, maybeOptions) {

	var targetScriptUrl = scriptUrl;
	var options = maybeOptions;

	if (typeof mapScriptUrlOrOptions === 'function') {
		targetScriptUrl = mapScriptUrlOrOptions(targetScriptUrl);
	} else {
		options = mapScriptUrlOrOptions;
	}

	if ('serviceWorker' in navigator) {
		return navigator.serviceWorker.register(targetScriptUrl, options);
	}

	return Promise.reject(new ServiceWorkerNoSupportError());
}
"
`;

exports[`service-worker-loader should emit bundle and service worker 2`] = `
"/* eslint-env serviceworker */
/* eslint-disable import/unambiguous */

self.addEventListener('install', () => {
	postMessageToAll({
		action: 'install'
	});
});

self.addEventListener('activate', () => {
	postMessageToAll({
		action: 'activate'
	});
});

self.addEventListener('fetch', (event) => {
	postMessageToAll({
		action:  'fetch',
		payload: event.request.url
	});
});

async function postMessageToAll(message) {

	const clients = await self.clients.matchAll();

	return Promise.all(
		clients.map(client => client.postMessage(message))
	);
}
"
`;
